Special handling AssignmentExpression with a MemberExpression on the right node (fixes #124)

rpl opened the pull request:
In this pull requests there are:
- a couple of additional test cases to explicitly cover the issue describe in #124 
- a draft for a proposed approach to deal with this scenario (to be evaluated and discussed as part of this pull request):
  - while validating a CallExpression, if we do have an AssignmentExpression as the callee node 
  - check if the AssignmentExpression has a MemberExpression on the right and, if it does, validate a fake call expression that does have the memberExpression as the callee node.

mozfreddyb commented on tests/rules/method.js:
```diff
@@ -201,6 +201,10 @@ eslintTester.run("method", rule, {
             parserOptions: { ecmaVersion: 6 },
+        },
+        { // Regression test for #124
+            code: "(e = this.n[n.i])(i, r)",
```
Can you expand by using an example that calls an "unsafe" function but with a safe parameter, to ensure that the tested code goes deeper? E.g. `"(e = node.insertAdjacentHtml('afterend', '<s>hello</s>')"`

rpl replied:
:+1: I added an additional case for that scenario.

mozfreddyb commented on tests/rules/method.js:
```diff
@@ -456,5 +460,15 @@ eslintTester.run("method", rule, {
             ]
         },
+        { // Regression test for #124
+            code: "(e = node.insertAdjacentHTML)(i, r)",
```
I realize I'm a bit pedantic, but please rename `r` to `evil` for a bit of extra clarity

rpl replied:
nah, it is not pedantic, I should have named it `evil` for both clarity and consistency.
Test case updated as suggested.

rpl commented:
@mozfreddyb I just updated this pull request based on the review comments and also rebased it on top of the current master to resolve the merging conflicts due to the changes in the test file.
This should now be ready to go.

mozfreddyb commented on tests/rules/method.js:
```diff
@@ -494,6 +502,26 @@ eslintTester.run("method", rule, {
+            ]
+        },
+        { // Regression test for #124, make sure we go deeper and detect the unsafe pattern
+            code: "(e = node.insertAdjacentHTML('beforebegin', evil))()",
```
Thinking about this, shouldn't this give us a Syntax Tree that, when traversed, looks no different than `node.insertAdjacentHTML('beforebegin', evil)`? I'm not sure how we can test that the test adds something
Would this be possible with coverage? We can do that as a follow-up, I'm happy to merge this as it is.

mozfreddyb replied:
this is now #128
